omod/src/main/java/org/openmrs/web/controller/PortletController.java

```diff
@@ -195,43 +204,68 @@ public ModelAndView handleRequest(HttpServletRequest request, HttpServletRespons
 					}
 					
 					if (Context.hasPrivilege(PrivilegeConstants.GET_OBS)) {
-						List<Obs> patientObs = Context.getObsService().getObservationsByPerson(p);
-						model.put("patientObs", patientObs);
+						Integer limit = getLimitParameter(request, as, "dashboard.observations.maximumNumberToShow");
+						Integer startIndex = getStartIndexParameter(request);
+						
+						Person person = (Person) p;
+						List<Person> persons = Collections.singletonList(person);
+						
+						// Get most recent observations using limit parameter
+						List<Obs> paginatedObs = Context.getObsService().getObservations(persons, null, null, null, null,
+						    null, Collections.singletonList("obsDatetime desc"), limit, startIndex, null, null, false);
+						
+						model.put("limit", limit);
+						model.put("patientObs", paginatedObs);
+						
+						// Handle BMI calculation
 						Obs latestWeight = null;
 						Obs latestHeight = null;
 						String bmiAsString = "?";
+						
 						try {
 							String weightString = as.getGlobalProperty("concept.weight");
 							ConceptNumeric weightConcept = null;
 							if (StringUtils.hasLength(weightString)) {
 								weightConcept = cs.getConceptNumeric(GeneralUtils.getConcept(weightString).getConceptId());
 							}
+							
 							String heightString = as.getGlobalProperty("concept.height");
 							ConceptNumeric heightConcept = null;
 							if (StringUtils.hasLength(heightString)) {
 								heightConcept = cs.getConceptNumeric(GeneralUtils.getConcept(heightString).getConceptId());
 							}
-							for (Obs obs : patientObs) {
```

Comment:
Rather than loading all observations and finding the latest in memory, we would leverage the database to do this work more efficiently. Not directly related to pagination but aligns with the optimization goals since we no longer need to iterate over all patientObs to extract weight and height.

Reply:
With performance optimisations, unless you have tested and confirmed, you can be very shocked to find a different outcome. The database itself being an expensive resource, and you have anyway already fetched these observations in the above call, you can find that actually iterating through the collection in memory may be faster that making another expensive database call. So i would only change that code if i have done some confirmations.

Reply:
@dkayiwa The paginated list only contains a limited number of observations (default 50), therefore BMI calculation couldn't find the necessary values, resulting in "?" if at all we add back the for loop while paginating. Now for this to work I would have to do something like this

```java
// Paginated observations for display
List<Obs> paginatedObs = Context.getObsService().getObservations(..., limit, startIndex, ...);

// Additional query to get ALL observations for BMI calculation
List<Obs> allObs = Context.getObsService().getObservations(persons, null, null, null, null,
    null, null, null, null, null, null, false);

// Finding weight/height in complete set of observations
for (Obs obs : allObs) {
    // ... looking for weight/height
}
``` 

In essence, we'd now be using:
- Paginated query for display purposes (performance optimization)
- Complete query for BMI calculation (functionality requirement)

Does this sound an approach to use?

Reply:
What do you see as the best approach?

Reply:
I think I will go with this option I have suggested, in the long run we have separation of concerns. Change made